src/integer_mod_q/modulus_polynomial_ring_zq/from.rs

```diff
@@ -89,15 +234,156 @@ impl FromStr for ModulusPolynomialRingZq {
     /// let poly_mod = ModulusPolynomialRingZq::from_str("3  1 0 1 mod 17").unwrap();
     /// ```
     /// # Errors and Failures
-    /// - Returns a [`MathError`] of type [`MathError::StringConversionError`]
-    ///     - if the string was not formatted correctly, e.g. not a correctly
-    ///         formatted [`PolyOverZq`].
-    ///         For further information see [`PolyOverZq::from_str`].
+    /// - Returns a [`MathError`] of type
+    ///     [`StringConversionError`](MathError::StringConversionError)
+    ///     - if the provided first half of the string was not formatted correctly to
+    ///         create a [`PolyOverZ`],
+    ///     - if the provided second half of the
+    ///         string was not formatted correctly to create a [`Modulus`],
+    ///     - if the number of coefficients was smaller than the number provided
+    ///         at the start of the provided string, or
+    ///     - if the provided value did not contain two whitespaces.
+    ///     - For further information see [`PolyOverZq::from_str`].
     /// - Returns a [`MathError`] of type
     ///     [`InvalidModulus`](MathError::InvalidModulus)
-    ///     if `modulus` is smaller than `2`.
+    ///     - if `modulus` is smaller than `2`, or
+    ///     - if the modulus polynomial is `0` or `1`.
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(Self::from(&PolyOverZq::from_str(s)?))
+        let poly_zq = PolyOverZq::from_str(s)?;
+
+        check_poly_mod(&poly_zq)?;
+
+        Ok(Self::from(poly_zq))
+    }
+}
+
+/// Checks weather a given [`PolyOverZq`] can be used as a [`ModulusPolynomialRingZq`].
+///
+/// Parameters:
+/// - `poly_zq`: the [`PolyOverZq`] value that should be checked.
+///
+/// Returns an empty `Ok`, or an [`MathError`] if the polynomial is no valid modulus polynomial.
+///
+/// # Examples
+/// ```compile_fail
+/// use qfall_math::integer_mod_q::PolyOverZq;
+/// use std::str::FromStr;
+///
+/// let poly_zq = PolyOverZq::from_str("2  1 2 mod 17").unwrap();
+///
+/// check_poly_mod(&poly_zq)?
+/// ```
+///
+/// # Errors and Failures
+/// - Returns a [`MathError`] of type
+///     [`InvalidModulus`](MathError::InvalidModulus)
+///     if the modulus polynomial is `0` or `1`.
+pub(crate) fn check_poly_mod(poly_zq: &PolyOverZq) -> Result<(), MathError> {
```

Comment:
Why didn't you just use a `PolyOverZ` as input? You don't check the modulus anyway.

Furthermore, I don't know how a function behaves that has a negative modulus-polynomial like -1. I would believe that this wouldn't allow to host any information as well (just like 1 wouldn't allow it). Please speak to Laurens to check for all possible cases in this method before you merge it.

Reply:
We need a PolyOverZq for the conversion into a ModulusPolynomialRingZq and since the function is only used for conversions to that type, it is faster to insert the PolyOverZq into the function rather than transforming it into a PolyOverZ.

I spoke with Laurens about that and we agreed on only rejecting 0 polynomials and allowing both 1 and -1.